OpenRA.Mods.Common/Traits/BotModules/HarvesterBotModule.cs

```diff
@@ -99,6 +111,30 @@ void IBotTick.BotTick(IBot bot)
 				AIUtils.BotDebug("AI: Harvester {0} is idle. Ordering to {1} in search for new resources.".F(harvester, newSafeResourcePatch));
 				bot.QueueOrder(new Order("Harvest", harvester, Target.FromCell(world, newSafeResourcePatch), false));
 			}
+
+			// Less harvesters than refineries - build a new harvester
+			if (ShouldBuildHarvester())
+			{
+				var unitBuilder = requestUnitProduction.FirstOrDefault(Exts.IsTraitEnabled);
+				if (unitBuilder != null)
+				{
+					var harvInfo = AIUtils.GetInfoByCommonName(Info.HarvesterTypes, player);
+					if (unitBuilder.RequestedProductionCount(bot, harvInfo.Name) == 0)
+						unitBuilder.RequestUnitProduction(bot, harvInfo.Name);
+				}
+			}
+		}
+
+		bool ShouldBuildHarvester()
+		{
+			if (!Info.HarvesterTypes.Any())
+				return false;
+
+			if (AIUtils.CountBuildingByCommonName(Info.HarvesterFactoryTypes, player) == 0)
```

Comment:
If you check the `unitBuilder.RequestedProductionCount` here instead then we should be able to drop the factory types completely - it'll queue a single harv, then wait until that is completed before doing anything else.  This simplifies the yaml config, and also saves perf.

Reply:
Done and moved the whole check inline into BotTick.